### ISSUE TITLE
gen_udp:open/2 does not set {reuseaddr, true} on Windows 10

### ISSUE THREAD

**Issue by ziopio** [erlang/otp#4819]
**Describe the bug**
On Windows 10 64-bit, when i try to open an UDP port with option {reuseaddr, true} the option is not applied.
The resulting socket owns the port exclusively. This makes it impossible to open multiple sockets on the same port.
The bug **DOES NOT** occur on an OTP linux installation executing through WSL.

**To Reproduce** (OTP 23.3 Eshell V11.2,  Windows 10 64-bit )
`1>  {_,S} = gen_udp:open(0,[{reuseaddr, true}]).`
`{ok,#Port<0.4>}`
`2> inet:getopts(S,[reuseaddr]).`
`{ok,[{reuseaddr,false}]}`

**Expected behavior** (OTP 22 Eshell V10.6.4, Ubuntu 20.04 on WSL )
`1>  {_,S} = gen_udp:open(0,[{reuseaddr, true}]).`
`{ok,#Port<0.4>}`
`2> inet:getopts(S,[reuseaddr]).`
`{ok,[{reuseaddr,true}]}`

**Affected versions**
Erlang/OTP 23.3 with Eshell V 11.2, 64-bit binary installation on Windows 10

**Additional context**
I my case i am implementing the RTPS protocol by ObjectManagmentGroup in Erlang. The p2p discovery requires multiple participants to listen to the same port for UDP multicast packets. This bug prevents me to start more than one node on my system.

**Comment by bmk** [comment]
After some searching on the net, I found that this (REUSEADDR) seems to have issues on windows.
For instance Apache HTTPD only use it on non-windows.

I also found the following comment:

"I understand REUSE is popular under *nix because it prevents re-binding process restart errors
 for processes stuck in TIME_WAIT.
 Supposedly Windows does not have this issue of preventing socket binding for socket addr/port combinations in TIME_WAIT.
 Apparently there is little value to using REUSE under Windows except in cases where you are using multicast."

**Comment by juhlig** [comment]
The same applies to `gen_tcp` etc. From `inet_drv.c`, with my favorite explanatory comment of all times 😁:
https://github.com/erlang/otp/blob/d646fa37c52af10fdf20e5194bb2d2df9701790a/erts/emulator/drivers/common/inet_drv.c#L6635-L6637

**Comment by ziopio** [comment]
So... it's more like an intended feature then a bug😅

**Comment by bmk** [comment]
I think it should be "feature" in this case...

**Comment by Maria-12648430** [comment]
"Bjorn says", priceless 😄

**Comment by juhlig** [comment]
> I think it should be "feature" in this case...

I don't remember all the details, but IIRC reuseaddr means something very different (and dangerous, of course) on Windows...

**Comment by juhlig** [comment]
https://docs.microsoft.com/en-us/windows/win32/winsock/using-so-reuseaddr-and-so-exclusiveaddruse

**Comment by bjorng** [comment]
On Tue, May 25, 2021 at 7:26 PM Maria Scott ***@***.***>
wrote:

> "Bjorn says", priceless 😄
>
>
> I think that comment was written by Tony Rogvall (
https://github.com/tonyrog).

The backstory was that at the time the Erlang distribution had its own
driver and there was a separate port program for applications that wanted
to use TCP and UDP. Tony was working on a new driver called the inet_driver
that could serve both purposes.

At that time, I was working with the Windows 95/NT build of Erlang. In the
socket code for Windows, we had removed the SO_REUSEADDR option, because we
had noticed that it would cause unsafe reuse of listen ports.

When Tony showed me code for the inet_driver, I saw that it used
SO_REUSEADDR on all platforms and I told him that he would have to take it
out for Windows.

He did.

/Björn

-- 
Björn Gustavsson, Erlang/OTP, Ericsson AB
